java-checks/src/main/java/org/sonar/java/checks/CollectionMethodsWithLinearComplexityCheck.java

```diff
@@ -0,0 +1,142 @@
+  private static final String CONCURRENT_LINKED_QUEUE = "java.util.concurrent.ConcurrentLinkedQueue";
+  private static final String CONCURRENT_LINKED_DEQUE = "java.util.concurrent.ConcurrentLinkedDeque";
+
+  private static MethodMatcher collectionMethodMatcher() {
+    return MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("java.util.Collection"));
+  }
+
+  private static final Map<MethodMatcher, Set<String>> matcherActualTypeMap;
+  static {
+    ImmutableMap.Builder<MethodMatcher, Set<String>> builder = ImmutableMap.builder();
+
+    MethodMatcher collectionContains = collectionMethodMatcher().name("contains").addParameter(JAVA_LANG_OBJECT);
+    builder.put(collectionContains, ImmutableSet.of(ARRAY_LIST, LINKED_LIST, COW_ARRAY_LIST, COW_ARRAY_SET, CONCURRENT_LINKED_QUEUE, CONCURRENT_LINKED_DEQUE));
+
+    MethodMatcher collectionSize = collectionMethodMatcher().name("size").withoutParameter();
+    builder.put(collectionSize, ImmutableSet.of(CONCURRENT_LINKED_QUEUE, CONCURRENT_LINKED_DEQUE));
+
+    MethodMatcher collectionAdd = collectionMethodMatcher().name("add").addParameter(TypeCriteria.anyType());
+    builder.put(collectionAdd, ImmutableSet.of(COW_ARRAY_SET, COW_ARRAY_LIST));
+
+    MethodMatcher collectionRemove = collectionMethodMatcher().name("remove").addParameter(JAVA_LANG_OBJECT);
+    builder.put(collectionRemove, ImmutableSet.of(ARRAY_LIST, COW_ARRAY_SET, COW_ARRAY_LIST));
+
+    MethodMatcher listGet = MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("java.util.List")).name("get").addParameter("int");
+    builder.put(listGet, ImmutableSet.of(LINKED_LIST));
+    matcherActualTypeMap = builder.build();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree mit = (MethodInvocationTree) tree;
+    if (!invocationInMethod(mit)) {
+      return;
+    }
+    matcherActualTypeMap.forEach((methodMatcher, actualTypes) -> {
+      if (methodMatcher.matches(mit)) {
+        Symbol target = invocationTarget(mit);
+        if (target != null && isField(target) && matchesActualType(target, actualTypes)) {
+          IdentifierTree methodName = MethodsHelper.methodName(mit);
+          reportIssue(methodName, "This call to \"" + methodName.name() + "()\" may be a performance hot spot if the collection is large.");
+        }
+      }
+    });
+  }
+
+  private static boolean invocationInMethod(MethodInvocationTree mit) {
+    Tree parent = mit.parent();
+    while (parent != null && !parent.is(Tree.Kind.METHOD)) {
+      parent = parent.parent();
+    }
+    return parent != null;
+  }
+
+  private static boolean isField(Symbol symbol) {
+    return symbol.isVariableSymbol() && symbol.owner().isTypeSymbol() && !"this".equals(symbol.name()) && !"super".equals(symbol.name());
```

Comment:
Do you fully cover these cases in your test file? Can't see any `this.contains(...)`, or `super.get(...)`.